### PR TITLE
[MLIR][OpenMP]Add order-modifier support to Order clause

### PR DESCRIPTION
**Pull request by harishch4**
This adds order-modifier (reproducible|unconstrained) support to Order clause.

### CODE REVIEW

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp [hunk 1]
```diff
@@ -433,6 +433,47 @@ static void printScheduleClause(OpAsmPrinter &p, Operation *op,
     p << ", simd";
 }
 
+//===----------------------------------------------------------------------===//
+// Parser and printer for Order Clause
+//===----------------------------------------------------------------------===//
+
+// order ::= `order` `(` [order-modiﬁer ':'] concurrent `)`
```
Comment (shraiysh) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:440]: The grammar seems inaccurate. In the testcases it seems that the second value can be both `default` and `concurrent`. Can you check this once?

**Review by tblah — COMMENTED** [review]
The changes here look reasonable to me. Please wait for another reviewer to take a look before merging.

Have you seen that the clause representation is being changed in https://github.com/llvm/llvm-project/pull/92521?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 2]
```diff
@@ -459,6 +459,17 @@ def LoopNestOp : OpenMP_Op<"loop_nest", [SameVariadicOperandSize,
...
+def OMP_OrderModUnconstrained      : I32EnumAttrCase<"unconstrained", 1>;
+def OrderModifier
+    : I32EnumAttr<"OrderModifier", "OpenMP Order Modifier",
+                  [OMP_OrderModReproducible, OMP_OrderModUnconstrained]> {
+  let genSpecializedAttr = 0;
+  let cppNamespace = "::mlir::omp";
+}
+def OrderModifierAttr : EnumAttr<OpenMP_Dialect, OrderModifier,
+                                    "order_mod">;
+
```
Comment (skatrak) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:472]: I think this should go into OpenMPEnums.td.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp [hunk 3]
```diff
@@ -433,6 +433,47 @@ static void printScheduleClause(OpAsmPrinter &p, Operation *op,
...
+// order-modiﬁer ::= reproducible | unconstrained
+static ParseResult parseOrderClause(OpAsmParser &parser,
+                                    ClauseOrderKindAttr &kindAttr,
+                                    OrderModifierAttr &modifierAttr) {
+  using OrderKindT = decltype(std::declval<ClauseOrderKindAttr>().getValue());
+  using OrderModifierT = decltype(std::declval<OrderModifierAttr>().getValue());
```
Comment (skatrak) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:446]: What's the purpose of these aliases? I can't see any uses.

**Review by skatrak — APPROVED** [review]
Thank you, LGTM. Just a couple of nits from me.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp (hunk 3 above)
Comment (harishch4) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp]: Thanks for reporting. I removed them now.

**Comment by harishch4** [comment]
If there are no objections, I would like to proceed with merging this code.

**Review by skatrak — APPROVED** [review]
I don't have any further comments, thank you!

**Pull request merged**